homeassistant/components/zoneminder/switch.py

```diff
@@ -1,55 +1,87 @@
 """Support for ZoneMinder switches."""
 import logging
+from typing import Callable, List, Optional
 
 import voluptuous as vol
-from zoneminder.monitor import MonitorState
+from zoneminder.monitor import Monitor, MonitorState
 
-from homeassistant.components.switch import PLATFORM_SCHEMA, SwitchEntity
+from homeassistant.components.switch import (
+    DOMAIN as SWITCH_DOMAIN,
+    PLATFORM_SCHEMA,
+    SwitchEntity,
+)
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import CONF_COMMAND_OFF, CONF_COMMAND_ON
-import homeassistant.helpers.config_validation as cv
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import Entity
 
-from . import DOMAIN as ZONEMINDER_DOMAIN
+from .common import get_client_from_data, get_platform_configs
 
 _LOGGER = logging.getLogger(__name__)
 
+MONITOR_STATES = {
+    MonitorState[name].value: MonitorState[name]
+    for name in dir(MonitorState)
+    if not name.startswith("_")
+}
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {
-        vol.Required(CONF_COMMAND_ON): cv.string,
-        vol.Required(CONF_COMMAND_OFF): cv.string,
+        vol.Required(CONF_COMMAND_ON): vol.All(vol.In(MONITOR_STATES.keys())),
+        vol.Required(CONF_COMMAND_OFF): vol.All(vol.In(MONITOR_STATES.keys())),
     }
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the ZoneMinder switch platform."""
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity], Optional[bool]], None],
+) -> None:
+    """Set up the sensor config entry."""
+    zm_client = get_client_from_data(hass, config_entry.unique_id)
+    monitors = await hass.async_add_job(zm_client.get_monitors)
 
-    on_state = MonitorState(config.get(CONF_COMMAND_ON))
-    off_state = MonitorState(config.get(CONF_COMMAND_OFF))
+    if not monitors:
+        _LOGGER.warning("Could not fetch monitors from ZoneMinder")
+        return
 
     switches = []
-    for zm_client in hass.data[ZONEMINDER_DOMAIN].values():
-        monitors = zm_client.get_monitors()
-        if not monitors:
-            _LOGGER.warning("Could not fetch monitors from ZoneMinder")
-            return
+    for monitor in monitors:
+        for config in get_platform_configs(hass, SWITCH_DOMAIN):
+            on_state = MONITOR_STATES[config[CONF_COMMAND_ON]]
```

Comment:
Prefacing this with the fact that I'm not an expert on config flows / config entries.... so apologies if this is an obvious question. Once this data is stored in the config entry how does the user make changes to it?

Reply:
For that matter how does a user set this from the config UI when they are first setting up from the config UI (if they didn't have the configuration.yaml imported in?)

Reply:
Currently, HASS has no ability to change the initial options of a config entry. This is a forthcoming feature to HASS however.
For the switch platform, the user will have to configure this from yaml just like they did before. Upon HASS startup or creation of a new entry (via import or from UI), HASS will run the config entry through this platform's setup function.

Reply:
I've just finished a full read of all of the code in the PR and wanted to double check one other thing:

Since the unique_ id is generated from the hostname/monitorid and the on/off state names of the created switch, when the user makes a change to an on/off value in the yaml configuration, this will create a new switch entity (without removing the old one) because the unique_id will have changed (since we don't have the way for UI editing yet).

The user would then be able to simply delete the old entity via the UI right?